engine/mpm_solver.py

```diff
@@ -133,7 +135,7 @@ def __init__(
         else:
             indices = ti.ijk
 
-        offset = tuple(-self.grid_size // 2 for _ in range(self.dim))
+        offset = tuple((res[i] - self.grid_size) // 2 for i in range(self.dim))
```

Comment:
The old code was designed to let grid coordinate `[self.grid_size//2, self.grid_size//2]` be the actual origin `[0.0, 0.0]`. Maybe we can simply let the user pass in a larger `grid_size` such as 8192 (sparsity in Taichi is basically free) so that he or she gets enough grid cells, and we can keep this property? (Assuming you are modifying this because of access out of grid bounds.)

Reply:
Yes, it would be fine to restore this change.